torchtune/rlhf/loss/dpo.py

```diff
@@ -231,3 +231,91 @@ def forward(
+            reference model. Default is 0.1.
+        desirable_weight (float):  Desirable losses are weighed by this factor to counter
+            unequal number of desirable and undesirable paris. Default is 1.0.
+        undesirable_weight (float):  Undesirable losses are weighed by this factor to counter
+            unequal number of desirable and undesirable paris. Default is 1.0.
+    """
+
+    def __init__(
+        self,
+        beta: float = 0.1,
+        desirable_weight: float = 1.0,
+        undesirable_weight: float = 1.0,
+    ):
+        super().__init__()
+        self.beta = beta
+        self.desirable_weight = desirable_weight
+        self.undesirable_weight = undesirable_weight
+
+    def forward(
+        self,
+        policy_chosen_logps: torch.Tensor,
+        policy_rejected_logps: torch.Tensor,
+        reference_chosen_logps: torch.Tensor,
+        reference_rejected_logps: torch.Tensor,
+    ) -> Tuple[torch.Tensor, torch.Tensor, torch.Tensor]:
+        """
+        Compute the KTO loss for a batch of policy and reference model log probabilities.
+
+        Args:
+                    policy_chosen_logps (torch.Tensor): Log probabilities of the policy model
+                        for the chosen responses. Shape: (batch_size)
+                    policy_rejected_logps (torch.Tensor): Log probabilities of the policy model
+                        for the rejected responses. Shape: (batch_size)
+                    reference_chosen_logps (torch.Tensor): Log probabilities of the reference model
+                        for the chosen responses. Shape: (batch_size)
+                    reference_rejected_logps (torch.Tensor): Log probabilities of the reference model
+                        for the rejected responses. Shape: (batch_size)
+
+        Returns:
+                    Tuple[torch.Tensor, torch.Tensor, torch.Tensor]: A tuple of three tensors:
+                        - losses: The DPO loss for each example in the batch.
+                        - chosen_rewards: Rewards for the chosen responses.
+                        - rejected_rewards: Rewards for the rejected responses.
+
+        """
+
+        kl = torch.zeros(1).to(policy_chosen_logps.device)
+
+        chosen_logratios = policy_chosen_logps - reference_chosen_logps
+
+        chosen_losses = 1 - F.sigmoid(self.beta * (chosen_logratios - kl))
+
+        rejected_logratios = policy_rejected_logps - reference_rejected_logps
+
+        rejected_losses = 1 - F.sigmoid(self.beta * (kl - rejected_logratios))
+
+        losses = torch.cat(
+            (
+                self.desirable_weight * chosen_losses,
+                self.undesirable_weight * rejected_losses,
+            ),
+            0,
+        )
+
+        chosen_rewards = self.beta * (chosen_logratios).detach()
+        rejected_rewards = self.beta * (rejected_logratios).detach()
```

Comment:
I'm not convinced that this is correct. Could you share a reference implementation for this loss function? For example, in TRL, the KL calculation is done differently (https://github.com/huggingface/trl/blob/main/trl/trainer/kto_trainer.py#L1100) - you can see that `self.calculate_KL` is `True` for the KTO loss (https://github.com/huggingface/trl/blob/main/trl/trainer/kto_trainer.py#L545).

Reply:
This is the same in the official KTO repo https://github.com/ContextualAI/HALOs/blob/61b9ee6787c9f52c2f4578533866e1cb42f314ec/trainers.py#L836

Reply:
Yes, I saved this mostly as drafted(version without KL) will bring it

Reply:
Hmm, to actually get kl_logprobs we need to modify concatenated_forward

Reply:
I don't want to pass cfg in concatenated_forward

Reply:
We can actually ignore KL divergence and use APO variant of loss:

`rejected_losses = F.sigmoid(self.beta * rejected_logratios)`

Reply:
And we need to actually change structure of the batch